pyked/chemked.py

```diff
@@ -233,6 +241,238 @@ def get_dataframe(self, output_columns=None):
+        reference.set('preferredKey', citation)
+        reference.set('doi', self.reference.doi)
+
+        apparatus = etree.SubElement(root, 'apparatus')
+        kind = etree.SubElement(apparatus, 'kind')
+        kind.text = self.apparatus.kind
+
+        common_properties = etree.SubElement(root, 'commonProperties')
+        # ChemKED objects have no common properties once loaded. Check for properties
+        # among datapoints that tend to be common
+        common = []
+        composition = self.datapoints[0].composition
+        # Composition type *has* to be the same
+        composition_type = self.datapoints[0].composition_type
+        if all([composition == dp.composition for dp in self.datapoints]):
+            # initial composition is common
+            common.append('composition')
+            prop = etree.SubElement(common_properties, 'property')
+            prop.set('name', 'initial composition')
+
+            for species in composition:
+                component = etree.SubElement(prop, 'component')
+                species_link = etree.SubElement(component, 'speciesLink')
+                species_link.set('preferredKey', species['species-name'])
+                if species.get('InChI'):
+                    species_link.set('InChI', species['InChI'])
+
+                amount = etree.SubElement(component, 'amount')
+                amount.set('units', composition_type)
+                amount.text = str(species['amount'].magnitude)
+
+        # If multiple datapoints present, then find any common properties. If only
+        # one datapoint, then composition should be the only "common" property.
+        if len(self.datapoints) > 1:
+            for prop_name in datagroup_properties:
+                attribute = prop_name.replace(' ', '_')
+                quantity = getattr(self.datapoints[0], attribute)
+                if (quantity != None and
+                    all([quantity == getattr(dp, attribute) for dp in self.datapoints])
+                    ):
+                    common.append(prop_name)
+                    prop = etree.SubElement(common_properties, 'property')
+                    prop.set('description', '')
+                    prop.set('name', prop_name)
+                    prop.set('units', str(quantity.units))
+
+                    value = etree.SubElement(prop, 'value')
+                    value.text = str(quantity.magnitude)
+
+        # Ignition delay can't be common, unless only a single datapoint.
+
+        datagroup = etree.SubElement(root, 'dataGroup')
+        datagroup.set('id', 'dg1')
+        datagroup_link = etree.SubElement(datagroup, 'dataGroupLink')
+        datagroup_link.set('dataGroupID', '')
+        datagroup_link.set('dataPointID', '')
+
+        property_idx = {}
+        labels = {'temperature': 'T', 'pressure': 'P',
+                  'ignition delay': 'tau', 'pressure rise': 'dP/dt',
+                  }
+
+        for prop_name in datagroup_properties:
```

Comment:
I tried to simplify the next 40 lines or so, and also account for the fact that different datapoints can have different units for the quantity, so we have to convert, we can't assume all the units are the same as the first datapoint (if that datapoint even has the property we're interested in). The following can replace line 377 to line 419 (but doesn't account for my comment below on line 392):

```
        for prop_name in datagroup_properties:
            attribute = prop_name.replace(' ', '_')
            prop_indices = [i for i, dp in enumerate(self.datapoints)
                            if getattr(dp, attribute) is not None]
            if prop_name in common or not prop_indices:
                continue

            prop = etree.SubElement(datagroup, 'property')
            prop.set('description', '')
            prop.set('name', prop_name)
            units = str(getattr(self.datapoints[prop_indices[0]], attribute).units)
            prop.set('units', units)
            idx = 'x{}'.format(len(property_idx) + 1)
            property_idx[idx] = {'name': prop_name, 'units': units}
            prop.set('id', idx)
            prop.set('label', labels[prop_name])

        if 'composition' not in common:
            for species in self.datapoints[0].composition:
                prop = etree.SubElement(datagroup, 'property')
                prop.set('description', '')

                idx = 'x{}'.format(len(property_idx) + 1)
                property_idx[idx] = {'name': species['species-name']}
                prop.set('id', idx)
                prop.set('label', '[' + species['species-name'] + ']')
                prop.set('name', 'composition')
                prop.set('units', self.datapoints[0].composition_type)

                species_link = etree.SubElement(prop, 'speciesLink')
                species_link.set('preferredKey', species['species-name'])
                if species.get('InChI'):
                    species_link.set('InChI', species['InChI'])

        for dp in self.datapoints:
            datapoint = etree.SubElement(datagroup, 'dataPoint')
            for idx, val in property_idx.items():
                value = etree.SubElement(datapoint, idx)
                # handle regular properties a bit differently than composition
                if val['name'] in datagroup_properties:
                    quantity = getattr(dp, val['name'].replace(' ', '_')).to(val['units'])
                    value.text = str(quantity.magnitude)
                else:
                    # composition
                    for item in dp.composition:
                        if item['species-name'] == val['name']:
                            value.text = str(item['amount'].magnitude)
```